File: 0011_Container_With_Most_Water/Container_With_Most_Water.rb
# @param {Integer[]} height
# @return {Integer}
def max_area(height)
  right = height.length - 1
  return 0 if right < 1
  max = 0
  left = 0
  while left < right
    temp = 0
    if height[right] > height[left]
      temp = (right - left) * height[left]
      left += 1
    else
      temp = (right - left) * height[right]
      right -= 1
    end
    max = temp if max < temp
  end
  max
end

File: 0011_Container_With_Most_Water/Container_With_Most_Water.py
# -*- coding: utf-8 -*-

"""
Given n non-negative integers a1, a2, ..., an , where each represents a point at coordinate (i, ai). n vertical lines are drawn such that the two endpoints of line i is at (i, ai) and (i, 0). Find two lines, which together with x-axis forms a container, such that the container contains the most water.

Note: You may not slant the container and n is at least 2.

The above vertical lines are represented by array [1,8,6,2,5,4,8,3,7]. In this case, the max area of water (blue section) the container can contain is 49.

Example:

Input: [1,8,6,2,5,4,8,3,7]
Output: 49
"""


class Solution:
    def maxArea(self, height: List[int]) -> int:
        result, left, right = 0, 0, len(height) - 1
        while left < right:
            if height[left] < height[right]:
                temp = height[left] * (right - left)
                left += 1
            else:
                temp = height[right] * (right - left)
                right -= 1
            result = max(result, temp)
        return result
